src/v2/controllers/vdeployment_controller/trusted/rely_guarantee.rs

```diff
@@ -0,0 +1,159 @@
+            &&& req.obj.metadata.owner_references.is_Some()
+            &&& exists |vd: VDeploymentView| 
+                #[trigger] owner_references.contains(vd.controller_owner_ref())
+        }
+    }
+}
+
+// Update requests to vrs must carry a resource version on them
+// and cannot 1) update vrs owned by a VDeployment
+//         or 2) update vrs to become owned
+// by a VDeployment.
+pub open spec fn vd_rely_get_then_update_req(req: GetThenUpdateRequest) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        req.obj.kind == Kind::CustomResourceKind("vreplicaset"@) ==>
+            req.obj.metadata.resource_version.is_Some()
+            // Prevents 1): where other controllers update vrs already owned
+            // by a VDeployment.
+            && !{
+                let etcd_obj = s.resources()[req.key()];
+                let owner_references = etcd_obj.metadata.owner_references.get_Some_0();
+                &&& s.resources().contains_key(req.key())
+                &&& etcd_obj.metadata.resource_version.is_Some()
+                &&& etcd_obj.metadata.resource_version == req.obj.metadata.resource_version
+                &&& etcd_obj.metadata.owner_references.is_Some()
+                // Can pass the owner_references check
+                &&& owner_references.contains(req.owner_ref)
+                &&& exists |vd: VDeploymentView| 
+                    #[trigger] owner_references.contains(vd.controller_owner_ref())
+            }
+            // Prevents 2): where other controllers update vrs so they become
+            // owned by a VDeployment.
+            && (req.obj.metadata.owner_references.is_Some() ==>
+                    forall |vd: VDeploymentView| 
+                        ! #[trigger] req.obj.metadata.owner_references.get_Some_0().contains(vd.controller_owner_ref()))
+    }
+}
+
+// TODO: double check if only vrs can send update status req to itself
+//       the update to VRS's guarantee to support this should also be updated
+// similar to update requests, minus the condition on owner_references.
+// Q: how to get controller type in ClusterState?
+pub open spec fn vd_rely_update_status_req(req: UpdateStatusRequest) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        req.obj.kind == Kind::CustomResourceKind("vdeployment"@) ==> 
+            req.obj.metadata.resource_version.is_Some()
+            && !{
+                let etcd_obj = s.resources()[req.key()];
+                &&& s.resources().contains_key(req.key())
+                &&& etcd_obj.metadata.resource_version.is_Some()
+                &&& etcd_obj.metadata.resource_version == req.obj.metadata.resource_version
+            }
+    }
+}
+
+// Other controllers don't try to delete vrs owned by a VDeployment.
+pub open spec fn vd_rely_get_then_delete_req(req: GetThenDeleteRequest) -> StatePred<ClusterState> {
+    |s: ClusterState| {
+        req.key.kind == Kind::CustomResourceKind("vreplicaset"@) ==>
```

Comment:
Nit: Can we import the VReplicaSet kind and do `VReplicaSetView::kind()` instead?

Reply:
+1